src/bin/cargo/main.rs

```diff
@@ -52,18 +53,32 @@ fn builtin_aliases_execs(cmd: &str) -> Option<&(&str, &str, &str)> {
     BUILTIN_ALIASES.iter().find(|alias| alias.0 == cmd)
 }
 
+/// Resolve the aliased command from the [`Config`] with a given command string.
+///
+/// The search fallback chain is:
+///
+/// 1. Get the aliased command as a string.
+/// 2. If the aliased command is not a string, get it as an array.
+/// 3. If cannot find any, finds one insides [`BUILTIN_ALIASES`].
+/// 4. If still no command is found, return `None`.
 fn aliased_command(config: &Config, command: &str) -> CargoResult<Option<Vec<String>>> {
-    let alias_name = format!("alias.{}", command);
-    let user_alias = match config.get_string(&alias_name) {
-        Ok(Some(record)) => Some(
+    let alias_name = format!("alias.{command}");
+    let user_alias = match config.get::<Value<String>>(&alias_name) {
+        Ok(record) => Some(
             record
                 .val
                 .split_whitespace()
                 .map(|s| s.to_string())
                 .collect(),
         ),
-        Ok(None) => None,
-        Err(_) => config.get::<Option<Vec<String>>>(&alias_name)?,
+        Err(e) => match e.downcast_ref::<ConfigError>().map(|e| e.kind()) {
+            // Just report we cannot find anything.
+            Some(ConfigErrorKind::MissingKey) => None,
+            // Alias might be set as an toml array, fallback and try it.
+            Some(ConfigErrorKind::UnexpectedValue) => config.get::<Option<_>>(&alias_name)?,
+            // A unrecoverable error, e.g. TOML parsing error. Relay it to the user.
+            Some(ConfigErrorKind::Custom) | None => return Err(e),
+        },
     };
```

Comment:
Is the test from "test(alias): propagate parse error to user " related to this?  Being in an earlier commit, I would expect that test to be showing the behavior as it is today but it wasn't updated in this commit.

btw having a commit with a test for the old behavior and then the new behavior is great for clarifying what is being fixed, like for me in catching up with this issue.  I did similar in #11062.

Reply:
This is my own practice that writing expected behaviour first and let tests fail. It is not ideal as it makes the commit fail to pass all tests. I am happy to adapt to the workflow as yours!

Reply:
imo every commit should be valid; this wreaks havoc with anyone doing bisects.  I wish CI systems allowed for testing intermediate commits.  This is one of the reasons some people do squash commits which I personally don't like due to the loss of information.

Reply:
Just update the first commit to exercise old behaviour and add a new commit for new behavior. Still not every commit is valid for all tests. I am uncertain if we should include test update in the same commit of behavior fix. (That is, merging f357db1 and bdd9ad3 in this PR).

Reply:
imo they should be; the commits aren't atomic as-is